src/backend/distributed/operations/citus_tools.c

```diff
@@ -93,6 +101,107 @@ CheckConnectionToNode(char *nodeName, uint32 nodePort)
 }
 
 
+/*
+ * citus_check_cluster_node_health UDF performs connectivity checks from all the nodes to
+ * all the nodes, and report success status
+ */
+Datum
+citus_check_cluster_node_health(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	TupleDesc tupleDescriptor = NULL;
+	Tuplestorestate *tupleStore = SetupTuplestore(fcinfo, &tupleDescriptor);
+
+	StoreAllConnectivityChecks(tupleStore, tupleDescriptor);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * GetConnectivityCheckCommand returns the command to check connections to a node
+ */
+static char *
+GetConnectivityCheckCommand(const char *nodeName, const uint32 nodePort)
+{
+	StringInfo connectivityCheckCommand = makeStringInfo();
+	appendStringInfo(connectivityCheckCommand,
+					 "SELECT citus_check_connection_to_node('%s', %d)",
+					 nodeName, nodePort);
+
+	return connectivityCheckCommand->data;
+}
+
+
+/*
+ * StoreAllConnectivityChecks performs connectivity checks from all the nodes to all the
+ * nodes, and report success status.
+ *
+ * Algorithm is:
+ * for sourceNode in activePrimaryWorkerList:
+ *   c = connectToNode(sourceNode)
+ *   for targetNode in activePrimaryWorkerList:
+ *     status = c.execute("SELECT citus_check_connection_to_node(targetNode.name, targetNode.port")
+ *     emit sourceNode.name, sourceNode.port, targetNode.name, targetNode.port, status
+ */
+static void
+StoreAllConnectivityChecks(Tuplestorestate *tupleStore, TupleDesc tupleDescriptor)
+{
+	Datum values[CONNECTIVITY_CHECK_COLUMNS];
+	bool isNulls[CONNECTIVITY_CHECK_COLUMNS];
+
+	List *workerNodeList = ActivePrimaryNodeList(ShareLock);
+
+	/* we want to check for connectivity in a deterministic order */
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodes);
+
+	WorkerNode *sourceWorkerNode = NULL;
+	foreach_ptr(sourceWorkerNode, workerNodeList)
+	{
+		const char *sourceNodeName = sourceWorkerNode->workerName;
+		const int sourceNodePort = sourceWorkerNode->workerPort;
+		int32 connectionFlags = 0;
+
+		MultiConnection *connectionToSourceNode =
+			GetNodeConnection(connectionFlags, sourceNodeName, sourceNodePort);
+
+		WorkerNode *targetWorkerNode = NULL;
+		foreach_ptr(targetWorkerNode, workerNodeList)
+		{
+			const char *targetNodeName = targetWorkerNode->workerName;
+			const int targetNodePort = targetWorkerNode->workerPort;
+
+			char *connectivityCheckCommandToTargetNode =
```

Comment:
There is one topic that I have been thinking about this. This gives the connection pairs from the coordinator's perspective of the cluster (e.g., using coordinator's metadatta). In reality, it should have been per-node's perspective of the cluster.

But, given that all nodes should have the same metadata with the current architecture, we should be fine.

Do you agree?

Reply:
I agree. The coordinator should decide on the node pairs. If one of the worker nodes is missing some `pg_dist_node` records, we may end up skipping some of the checks, and potentially give a false success message when we run:

	SELECT bool_and(success) FROM citus_check_cluster_node_health();